diff --git a/libavformat/flvdec.c b/libavformat/flvdec.c
index b531a39adc..6b0a104f18 100644
--- a/libavformat/flvdec.c
+++ b/libavformat/flvdec.c
@@ -1056,7 +1056,7 @@ retry:
             int type;
             meta_pos = avio_tell(s->pb);
             type = flv_read_metabody(s, next);
-            if (type == 0 && dts == 0 || type < 0) {
+            if (type == 0 && dts == 0 || type < 0 || type == TYPE_UNKNOWN) {
                 if (type < 0 && flv->validate_count &&
                     flv->validate_index[0].pos     > next &&
                     flv->validate_index[0].pos - 4 < next
