--- a/qtbase/cmake/QtPublicDependencyHelpers.cmake
+++ b/qtbase/cmake/QtPublicDependencyHelpers.cmake
@@ -212,7 +212,13 @@ function(_qt_internal_determine_if_host_info_package_needed out_var)
 endfunction()
 
 macro(_qt_internal_find_host_info_package platform_requires_host_info)
-    if(${platform_requires_host_info})
+    if(DEFINED QT_REQUIRE_HOST_PATH_CHECK)
+        set(_qt_platform_requires_host_info "${QT_REQUIRE_HOST_PATH_CHECK}")
+    else()
+        set(_qt_platform_requires_host_info "${platform_requires_host_info}")
+    endif()
+
+    if(_qt_platform_requires_host_info)
         find_package(Qt6HostInfo
                      CONFIG
                      REQUIRED
