--- /dev/null	2021-11-07 15:50:49.000000000 +0100
+++ ./cross_compile.txt		2021-11-07 15:50:00.000000000 +0100
@@ -0,0 +1,20 @@
+[host_machine]
+system = 'darwin'
+cpu_family = 'aarch64'
+cpu = 'arm64'
+endian = 'little'
+
+[binaries]
+c = 'cc'
+cpp = 'c++'
+ar = 'ar'
+strip = 'strip'
+pkgconfig = 'pkg-config'
+
+[built-in options]
+c_std = 'c11'
+cpp_std = 'c++11'
+c_args = '-target arm64-apple-macos11 -mmacosx-version-min=11.0'
+cpp_args = '-target arm64-apple-macos11 -mmacosx-version-min=11.0'
+c_link_args = '-target arm64-apple-macos11 -mmacosx-version-min=11.0'
+cpp_link_args = '-target arm64-apple-macos11 -mmacosx-version-min=11.0'
--- ./include/meson.build	2021-07-22 09:28:00.000000000 +0200
+++ ./include/meson.build	2021-11-10 17:23:06.000000000 +0100
@@ -2,14 +2,11 @@
 # SPDX-License-Identifier: BSD-2-Clause
 
 #stub file
-librist_git_dir = join_paths(librist_src_root, '.git')
-rev_target = vcs_tag(command: [
-        'git', '--git-dir', librist_git_dir,
-		'--work-tree', meson.source_root(),
-        'describe', '--tags', '--dirty',
-        '--match', 'v?.*', '--always'
-    ],
+version_data = configuration_data()
+version_data.set('VCS_TAG', '0.2.6')
+rev_target = configure_file(
     input: 'vcs_version.h.in',
-    output: 'vcs_version.h'
+    output: 'vcs_version.h',
+    configuration: version_data
 )
 subdir('librist')
