diff --git a/CMakeLists.txt b/CMakeLists.txt
index ed6e603..6931ff3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -59,14 +59,6 @@ endif()
 # Set CMAKE_BUILD_TYPE properly, now that you know
 # that ENABLE_DEBUG is set as it should.
 
-if (ENABLE_DEBUG EQUAL 2)
-	set (CMAKE_BUILD_TYPE "RelWithDebInfo")
-elseif (ENABLE_DEBUG) # 1, ON, YES, TRUE, Y, or any other non-zero number
-	set (CMAKE_BUILD_TYPE "Debug")
-else()
-	set (CMAKE_BUILD_TYPE "Release")
-endif()
-
 message(STATUS "BUILD TYPE: ${CMAKE_BUILD_TYPE}")
 
 getVarsWith(ENFORCE_ enforcers)
diff --git a/srt-ffplay b/srt-ffplay
deleted file mode 120000
index d7b96ee..0000000
--- a/srt-ffplay
+++ /dev/null
@@ -1 +0,0 @@
-scripts/srt-ffplay
\ No newline at end of file
