diff --git a/cmake/Modules/FindMbedTLS.cmake b/cmake/Modules/FindMbedTLS.cmake
index f7b5618..0b11208 100644
--- a/cmake/Modules/FindMbedTLS.cmake
+++ b/cmake/Modules/FindMbedTLS.cmake
@@ -84,7 +84,7 @@ macro(MbedTLS_set_soname component)
     endif()
   elseif(CMAKE_HOST_SYSTEM_NAME MATCHES "Linux|FreeBSD")
     execute_process(
-      COMMAND sh -c "objdump -p '${Mbed${component}_LIBRARY}' | grep SONAME"
+      COMMAND sh -c "${CMAKE_OBJDUMP}  -p '${Mbed${component}_LIBRARY}' | grep SONAME"
       OUTPUT_VARIABLE _output
       RESULT_VARIABLE _result)
 
-- 
2.39.1.windows.1

