diff --git a/include/mbedtls/config.h b/include/mbedtls/config.h
index 1e6e05275..98b5be699 100644
--- a/include/mbedtls/config.h
+++ b/include/mbedtls/config.h
@@ -1971,7 +1971,7 @@
  *
  * Uncomment this to allow your own alternate threading implementation.
  */
-//#define MBEDTLS_THREADING_ALT
+#define MBEDTLS_THREADING_ALT
 
 /**
  * \def MBEDTLS_THREADING_PTHREAD
@@ -3286,7 +3286,7 @@
  *
  * Enable this layer to allow use of mutexes within mbed TLS
  */
-//#define MBEDTLS_THREADING_C
+#define MBEDTLS_THREADING_C
 
 /**
  * \def MBEDTLS_TIMING_C
