Submodule qtbase contains modified content
diff --git a/qtbase/src/widgets/styles/qwindowsstyle.cpp b/qtbase/src/widgets/styles/qwindowsstyle.cpp
index 39fabd1218..a2b4cc0ccc 100644
--- a/qtbase/src/widgets/styles/qwindowsstyle.cpp
+++ b/qtbase/src/widgets/styles/qwindowsstyle.cpp
@@ -409,7 +409,10 @@ static QScreen *screenOf(const QWidget *w)
 // and account for secondary screens with differing logical DPI.
 qreal QWindowsStylePrivate::nativeMetricScaleFactor(const QWidget *widget)
 {
-    qreal result = qreal(1) / QWindowsStylePrivate::devicePixelRatio(widget);
+    const QPlatformScreen *screen = screenOf(widget)->handle();
+    const qreal scale = screen ? (screen->logicalDpi().first / screen->logicalBaseDpi().first)
+                               : QWindowsStylePrivate::appDevicePixelRatio();
+    qreal result = qreal(1) / scale;
     if (QGuiApplicationPrivate::screen_list.size() > 1) {
         const QScreen *primaryScreen = QGuiApplication::primaryScreen();
         const QScreen *screen = screenOf(widget);
