diff --git a/libavformat/hlsenc.c b/libavformat/hlsenc.c
index 7c37bc50b9..ad3800ecae 100644
--- a/libavformat/hlsenc.c
+++ b/libavformat/hlsenc.c
@@ -316,8 +316,7 @@ static int hlsenc_io_close(AVFormatContext *s, AVIOContext **pb, char *filename)
         URLContext *http_url_context = ffio_geturlcontext(*pb);
         av_assert0(http_url_context);
         avio_flush(*pb);
-        ffurl_shutdown(http_url_context, AVIO_FLAG_WRITE);
-        ret = ff_http_get_shutdown_status(http_url_context);
+        ret = ffurl_shutdown(http_url_context, AVIO_FLAG_WRITE);
 #endif
     }
     return ret;
