--- ./libavformat/flvdec.c	2021-10-24 22:47:07.000000000 +0200
+++ ./libavformat/flvdec.c	2021-11-08 13:13:47.000000000 +0100
@@ -1077,7 +1077,7 @@
             int type;
             meta_pos = avio_tell(s->pb);
             type = flv_read_metabody(s, next);
-            if (type == 0 && dts == 0 || type < 0) {
+            if (type == 0 && dts == 0 || type < 0 || type == TYPE_UNKNOWN) {
                 if (type < 0 && flv->validate_count &&
                     flv->validate_index[0].pos     > next &&
                     flv->validate_index[0].pos - 4 < next) {
--- ./libavcodec/libaomenc.c  2021-10-24 22:47:07.000000000 +0200
+++ ./libavcodec/libaomenc.c  2021-11-08 13:15:54.000000000 +0100
@@ -1337,7 +1337,7 @@
 #define OFFSET(x) offsetof(AOMContext, x)
 #define VE AV_OPT_FLAG_VIDEO_PARAM | AV_OPT_FLAG_ENCODING_PARAM
 static const AVOption options[] = {
-    { "cpu-used",        "Quality/Speed ratio modifier",           OFFSET(cpu_used),        AV_OPT_TYPE_INT, {.i64 = 1}, 0, 8, VE},
+    { "cpu-used",        "Quality/Speed ratio modifier",           OFFSET(cpu_used),        AV_OPT_TYPE_INT, {.i64 = 1}, 0, 10, VE},
     { "auto-alt-ref",    "Enable use of alternate reference "
                          "frames (2-pass only)",                   OFFSET(auto_alt_ref),    AV_OPT_TYPE_INT, {.i64 = -1},      -1,      2,       VE},
     { "lag-in-frames",   "Number of frames to look ahead at for "
