--- ./qtbase/mkspecs/common/mac.conf    2022-01-10 20:39:33.000000000 +0100
+++ ./qtbase/mkspecs/common/mac.conf  2022-01-10 20:37:19.000000000 +0100
@@ -7,7 +7,7 @@
 QMAKE_PLATFORM         += mac darwin
 
 MAKEFILE_GENERATOR      = UNIX
-CONFIG                 += app_bundle incremental global_init_link_order lib_version_first sdk
+CONFIG                 += app_bundle incremental global_init_link_order lib_version_first sdk silent
 QMAKE_INCREMENTAL_STYLE = sublib
 
 include(unix.conf)
--- ./qtbase/mkspecs/common/macx.conf   2022-01-10 20:39:44.000000000 +0100
+++ ./qtbase/mkspecs/common/macx.conf 2022-01-10 20:29:39.000000000 +0100
@@ -12,7 +12,7 @@
 # older SDKs we have to keep this.
 QT_MAC_SDK_VERSION_MIN = 10.14
 
-QT_MAC_SDK_VERSION_MAX = 11.0
+QT_MAC_SDK_VERSION_MAX = 13.0
 
 device.sdk = macosx
 device.target = device
