--- ./qmake/Makefile   2021-06-12 15:29:37.000000000 +0200
+++ ./qmake/Makefile    2021-06-12 15:35:54.000000000 +0200
@@ -38,8 +38,8 @@
 CARBON_LFLAGS = -framework ApplicationServices
 CARBON_CFLAGS = -fconstant-cfstrings
 
-EXTRA_CXXFLAGS = $(CARBON_CFLAGS)
-EXTRA_LFLAGS = $(COCOA_LFLAGS) $(CARBON_LFLAGS)
+EXTRA_CXXFLAGS += $(CARBON_CFLAGS)
+EXTRA_LFLAGS += $(COCOA_LFLAGS) $(CARBON_LFLAGS)
 
 QTOBJS2 = \
     qsettings_mac.o \
